### DISCUSSION TITLE
Am I able to use some EJBCA  Enterprise functionalities with EJBCA Community edition ?

### DISCUSSION THREAD

**Discussion by fco-code-org** [Keyfactor/ejbca-ce#423] [General]
Hello everyone,

I want to perform a POC on EST. But EST needs EJBCA Enterprise.
Is it still possible to use some EJBCA Enterprise functionalities for testing purpose with EJBCA CE ?

Best Regards.

**Comment by primetomas** [comment]
You can test EJBCA Enterprise Cloud in AWS or Azure, for free for 30 days. https://www.ejbca.org/download/

**Reply by fco-code-org** [reply]
Hello,

Thanks for you answer.

If we want to use EJBCA enterprise, the exhaustive options are :
- EJBCA SaaS
- EJBCA Cloud
- EJBCA Appliance
- EJBCA Software (eg software appliance)

Do you confirm that there is no other way to use EJBCA Enterprise ? 

Best Regards.

**Reply by primetomas** [reply]
There will be (is) also an EJBCA Container. Like the container with EJBCA on DockerHub but for Enterprise. And a software appliance.

So the complete list is:
- EJBCA SaaS
- EJBCA Cloud
- EJBCA Hardware Appliance
- EJBCA Software Appliance (VM)
- EJBCA Container
- EJBCA Software

**Reply by fco-code-org** [reply]
Hello,

Thanks again for your answer.

If I look at [https://doc.primekey.com/ejbca/tutorials-and-guides/tutorial-start-out-with-ejbca-docker-container](https://doc.primekey.com/ejbca/tutorials-and-guides/tutorial-start-out-with-ejbca-docker-container)  
It is writen "Learn how to use Docker with Docker Compose to run the EJBCA Community container with a MariaDB database container." So it is definitively not EJBCA Enterprise, right ?

What do you mean by EJBCA Software ? What's the differences between EJBCA Software and EJBCA Software Appliance (VM) ?

Best Regards.

**Reply by primetomas** [reply]
Yes, the tutorials are designed so you can use them with Community, since Enterprise is not publicly available.

EJBCA Software is a zip file with source code that you have to build, just like the source zip download of EJBCA Community from SourceForge. Software appliance is a pre-packaged VM with everything included, application server, database, EJBCA, and easy web configuration for clustering, HSM support, etc, making it easy to deploy, operate and upgrade. With "Software" you have to configure everything yourself, install a database, HSM drivers, etc, etc.

**Reply by primetomas** [reply]
You can find some information about Community vs Enterprise here: https://www.ejbca.org/community-vs-enterprise/

**Reply by fco-code-org** [reply]
Hello,

So EJBCA Software is a zip containing the source code with the Enterprise features in it, right ?
This one is only available on SourceForge right ?

Best Regards.

**Reply by primetomas** [reply]
The zip file with EJBCA Community is available on SourceForge, see here https://www.ejbca.org/download/
EJBCA Enterprise is only available from Keyfactor. https://www.keyfactor.com/products/ejbca-enterprise/

**Reply by fco-code-org** [reply]
Hello,

EJBCA Enterprise is only available from Keyfactor, fine ?
My understanding is that, if I want to use it (the Enterprise version), the options are limited:
- EJBCA SaaS
- EJBCA Cloud
- EJBCA Appliance
- EJBCA Software (eg software appliance)

There is no **EJBCA Enterprise**  delivered as a Docker images, right ?
I can't get the source code inluding the Enterprise features in it, to be able to build it on my own, right ? 

Best Regards.

**Reply by primetomas** [reply]
There _is_ a docker version of EJBCA Enterprise (not in the price list as of today, but will be soon). You can get the pure software to build on your own, but that is the most expensive option, and nothing that is offered as first choice.

If you want to try EJBCA Enterprise for free, EJBCA Cloud is the only option.

**Reply by fco-code-org** [reply]
Hello,

If we use the **EJBCA Appliance** or **EJBCA Software** how can we interact ?
As an EJBCA client could I use EJB calls to target these EJBCA flavors ? or do I have to call some REST APIs or SOAP APIs ?
Someone told me that the EJBCA Soap APIs are more complete than the REST ones: is it true ?

Best Regards.

**Reply by primetomas** [reply]
A Community client can call Enterprise and vice versa, no problem.

There are a few things you can do with the SOAP API that you can not do with REST (yet), but it is also true the other way around that REST has some functions that SOAP does not. If selecting between SOAP and REST I suggest REST as this is the API that will be enhanced as times goes on, SOAP is considered legacy and will not be updated with new functionality.

Then there are also the standard protocols of course, CMP or EST for example.

**Reply by fco-code-org** [reply]
Thanks for your replies :-)

Try to summarize:

- Your advice is from client perspective it should be better to use the EJBCA REST apis instead of the SOAP ones, right ?
- Not sure if you answered to this question: would a client be still able to use the EJB apis to target an EJBCA Software or an EJBCA Appliance  ?   


Another question, is possible from EJBCA Software or an EJBCA Appliance to target Thales DPoD (HSMaaS) ?
  
Merry Christmas and Happy new year for you and your family.

Best Regrads.

**Reply by primetomas** [reply]
EJB interfaces are not recommended to use from clients. That is only used by the local CLI on localhost of the server. EJB is not for remote client usage.
We recommend REST yes.
EJBCA Software and Software Appliance can use DPoD, but not Hardware Appliance because the HW Appliance comes with it's own HSM built in.

**Reply by fco-code-org** [reply]
Hello, 

You said: EJB is not for remote client usage.
Do you mean that EJB for remote client is not supported ?

Best Regards.

**Reply by primetomas** [reply]
Not sure if we speak about the same thing when you say EJB APIs. Let me rephrase.

Yes you can use the standard APIs and protocols in EJBCA against a software or hardware appliance. I.e. CMP, EST, REST, SOAP, SCEP.

**Reply by fco-code-org** [reply]
Hello,
Happy new year and best wishes for you and your family.
EJB APIs call means  call using EJB protocole like iiop, ....eg not SOAP or REST APIs

Best Regards.

**Reply by primetomas** [reply]
EJB API calls are not supported outside of localhost (WildFly by default only listens to localhost for it). Also the EJB API is an internal API from EJBCA point of view, and may change between every version of EJBCA.

**Reply by primetomas** [reply]
You can use it in plug-ins to EJBCA though, that's kind of the idea. But don't use it externally.
